contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1051 @@
+ *     - Used to keep track of each account's weighted ownership within a bucket
+ *     - Relative weight between buckets is meaningless
+ *     - Only accounts' relative weight within a bucket matters
+ *
+ * - Token Deposits:
+ *   - Go into a particular bucket, determined by time since the start of the position
+ *     - If the position has not started: bucket = 0
+ *     - If the position has started:     bucket = ceiling(time_since_start / BUCKET_TIME)
+ *     - This is always the highest bucket; that is, all higher buckets have no Weight, AA or OP
+ *   - Increase the bucket's AA
+ *   - Increase the bucket's weight and the account's weight in that bucket
+ *
+ * - Token Withdrawals:
+ *   - Can be from any bucket with available amount
+ *   - Decrease the bucket's AA (if it has enough, otherwise throw)
+ *   - Decrease the bucket's weight and the account's weight in that bucket
+ *
+ * - Increasing the Position (Lending):
+ *   - The lowest buckets with AA are used first
+ *   - Decreases AA
+ *   - Increases OP
+ *
+ * - Decreasing the Position (Being Paid-Back)
+ *   - The highest buckets with OP are paid back first
+ *   - Decreases OP
+ *   - Increases AA
+ *
+ *
+ * - Over time, this gives highest interest rates to earlier buckets, but disallows withdrawals from
+ *   those buckets for a longer period of time.
+ * - Deposits in the same bucket earn the same interest rate.
+ * - Lenders can withdraw their funds at any time if they are not being lent (and are therefore not
+ *   making the maximum interest).
+ * - The highest bucket with OP is always less-than-or-equal-to the lowest bucket with AA
+ */
+contract BucketLender is
+    OnlyMargin,
+    LoanOwner,
+    IncreaseLoanDelegator,
+    MarginCallDelegator,
+    CancelMarginCallDelegator,
+    ForceRecoverCollateralDelegator,
+    LoanOfferingVerifier,
+    ReentrancyGuard
+{
+    using SafeMath for uint256;
+
+    // ============ Events ============
+
+    event Deposit(
+        address beneficiary,
+        uint256 bucket,
+        uint256 amount,
+        uint256 weight
+    );
+
+    event Withdraw(
+        address withdrawer,
+        uint256 bucket,
+        uint256 weight,
+        uint256 owedTokenWithdrawn,
+        uint256 heldTokenWithdrawn
+    );
+
+    // ============ State Variables ============
+
+    /**
+     * Available Amount (AA) is the amount of tokens that is available to be lent by each bucket.
+     * These tokens are also available to be withdrawn by the accounts that have weight in the
+     * bucket.
+     */
+    // AA for each bucket
+    mapping(uint256 => uint256) public availableForBucket;
+    // Total AA
+    uint256 public availableTotal;
+
+    /**
+     * Outstanding Principal (OP) is the share of the margin position's principal that each bucket
+     * is responsible for. That is, each bucket with OP is owed (OP)*E^(RT) owedTokens in repayment.
+     */
+    // OP for each bucket
+    mapping(uint256 => uint256) public principalForBucket;
```

Comment:
it's confusing that the name of the variable doesn't match either of the things in the comment (Outstanding Principal / OP)

Reply:
Likewise for the available stuff too